### ISSUE TITLE
VB Double arithmetic is sensitive to floating point control word register

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#22533]
Create new VB Console App (.NET Framework) project with code below:

```VB
Imports System.Runtime.InteropServices

Class Module1
    <DllImport("msvcr120.dll", EntryPoint:="_controlfp_s", CallingConvention:=CallingConvention.Cdecl)>
    Public Shared Function ControlFP(ByRef currentCtrl As IntPtr, ByVal newCtrl As UInt32, ByVal mask As UInt32) As Int32
    End Function

    Public Shared Sub Main()

        Dim origFPCW, chgdFPCW As UInt32
        Dim errVal

        ' Verify initial state of register
        errVal = ControlFP(chgdFPCW, 0, 0)
        Console.WriteLine(If(chgdFPCW = &H9001F, "Register as expected", "NOT"))

        TestArithmetic() ' Prints FALSE

        Console.WriteLine("OFFICE component modifies register")
        chgdFPCW = 0
        origFPCW = &HC001F ' control word AFTER Office component invoked
        errVal = ControlFP(chgdFPCW, origFPCW, &HFFFFF)

        TestArithmetic() ' Prints TRUE

    End Sub

    Public Shared Sub TestArithmetic()
        Dim dValue As Double = 600.1
        Dim mbytDeciWgt As Byte = 1

        If CDbl(dValue) > CDbl(Int(dValue) + 10 ^ -mbytDeciWgt) Then
            Console.WriteLine("CDbl(dValue) > CDbl(Int(dValue) + 10 ^ -mbytDeciWgt) IS TRUE")
        Else
            Console.WriteLine("CDbl(dValue) > CDbl(Int(dValue) + 10 ^ -mbytDeciWgt) IS FALSE")
        End If
    End Sub
End Class
```
Note: the project builds for 32-bit (because the "Prefer 32-bit" option is selected by default in new projects).
Note: this can also repro with other C runtimes in the `DllImport` (not just `msvcr120.dll`).

VS2013 did not have this problem (it is resilient to the register being modified). It produces:
```
.method public static 
    void testVS2017 () cil managed 
{
    // Method begins at RVA 0x21cc
    // Code size 75 (0x4b)
    .maxstack 4
    .locals init (
        [0] float64 dValue,
        [1] uint8 mbytDeciWgt,
        [2] bool VB$CG$t_bool$S0
    )

    IL_0000: nop
    IL_0001: ldc.r8 600.1
    IL_000a: stloc.0
    IL_000b: ldc.i4.1
    IL_000c: stloc.1
    IL_000d: ldloc.0
    IL_000e: conv.r8 // <----- extra
    IL_000f: ldloc.0
    IL_0010: call float64 [Microsoft.VisualBasic]Microsoft.VisualBasic.Conversion::Int(float64)
    IL_0015: ldc.r8 10
    IL_001e: ldloc.1
    IL_001f: neg
    IL_0020: conv.ovf.i2
    IL_0021: conv.r8
    IL_0022: call float64 [mscorlib]System.Math::Pow(float64, float64)
    IL_0027: add
    IL_0028: conv.r8 // <----- extra
    IL_0029: cgt
    IL_002b: stloc.2
    IL_002c: ldloc.2
    IL_002d: brfalse.s IL_003c

    IL_002f: ldstr "CDbl(dValue) > CDbl(Int(dValue) + 10 ^ -mbytDeciWgt) IS TRUE"
    IL_0034: call void [mscorlib]System.Console::WriteLine(string)
    IL_0039: nop
    IL_003a: br.s IL_0048

    IL_003c: nop
    IL_003d: ldstr "CDbl(dValue) > CDbl(Int(dValue) + 10 ^ -mbytDeciWgt) IS FALSE"
    IL_0042: call void [mscorlib]System.Console::WriteLine(string)
    IL_0047: nop

    IL_0048: nop
    IL_0049: nop
    IL_004a: ret
} // end of method LiemTestVS2017::testVS2017
```

VS2017 produces:
```
.method public static 
	void testVS2017 () cil managed 
{
	// Method begins at RVA 0x2394
	// Code size 75 (0x4b)
	.maxstack 4
	.locals init (
		[0] float64 dValue,
		[1] uint8 mbytDeciWgt,
		[2] bool
	)

	IL_0000: nop
	IL_0001: ldc.r8 600.1
	IL_000a: stloc.0
	IL_000b: ldc.i4.1
	IL_000c: stloc.1
	IL_000d: ldloc.0
	IL_000e: ldloc.0
	IL_000f: call float64 [Microsoft.VisualBasic]Microsoft.VisualBasic.Conversion::Int(float64)
	IL_0014: ldc.r8 10
	IL_001d: ldloc.1
	IL_001e: neg
	IL_001f: conv.ovf.i2
	IL_0020: conv.r8
	IL_0021: call float64 [mscorlib]System.Math::Pow(float64, float64)
	IL_0026: add
	IL_0027: cgt
	IL_0029: stloc.2
	IL_002a: ldloc.2
	IL_002b: brfalse.s IL_003b

	IL_002d: ldstr "CDbl(dValue) > CDbl(Int(dValue) + 10 ^ -mbytDeciWgt) IS TRUE"
	IL_0032: call void [mscorlib]System.Console::WriteLine(string)
	IL_0037: nop
	IL_0038: nop
	IL_0039: br.s IL_0048

	IL_003b: nop
	IL_003c: ldstr "CDbl(dValue) > CDbl(Int(dValue) + 10 ^ -mbytDeciWgt) IS FALSE"
	IL_0041: call void [mscorlib]System.Console::WriteLine(string)
	IL_0046: nop
	IL_0047: nop

	IL_0048: br.s IL_004a

	IL_004a: ret
} // end of method LiemTestVS2017::testVS2017
```

**Comment by jcouv** [comment]
From discussion with Jared, the `conv.r8` is important (see his [blog post](http://blog.paranoidcoding.com/2014/12/22/redundant-cast.html) on the matter). It should be emitted whenever there is an explicit cast to double in the code. He remembers this exact problem was fixed in Roslyn for C#.

We need to confirm that the customer has explicit casts in their code. If so, a fix to the Roslyn VB compiler will fix their problem.

Then we'll want to update the [IDE issue for RemoveRedundantCast](https://github.com/dotnet/roslyn/issues/6309) to keep such casts in VB as well.

We'll also want to file an issue on the Office component that leaves the register in a different state than it found it.

**Comment by jcouv** [comment]
The fix is merged. It will first be available in Visual Studio 2017 version 15.5 preview 2.
